File: kotlin/lesson 7/IsValidBST.kt
// https://leetcode.com/problems/validate-binary-search-tree/

class Solution {
    fun isValidBST(root: TreeNode?): Boolean {
        return isValid(root)
    }
    
    private fun isValid(node: TreeNode?, min: Int? = null, max: Int? = null): Boolean {
        if (node == null) return true

        if (min != null && node.`val` <= min || max != null && node.`val` >= max) {
            return false
        }
        
        return isValid(node.left, min, node.`val`) && isValid(node.right, node.`val`, max)
    }
}

File: csharp/lesson 7/IsValidBST.cs
// https://leetcode.com/problems/validate-binary-search-tree/

public class Solution
{
    public bool IsValidBST(TreeNode root)
    {
        return IsValid(root);
    }
    
    private bool IsValid(TreeNode node, int? min = null, int? max = null)
    {
        if (node == null) return true;
        
        if (node.val <= min || node.val >= max) return false;
        
        return IsValid(node.left, min, node.val) && IsValid(node.right, node.val, max);
    }
}
